Dgraph Crashes Randomly While Running In GCE

```
panic: runtime error: index out of range

goroutine 244714 [running]:
github.com/dgraph-io/dgraph/algo.Difference(0xc44fa0e820, 0xc4534e41a0)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/algo/uidlist.go:242 +0x26d
github.com/dgraph-io/dgraph/query.ProcessGraph(0x1739f00, 0xc433e188d0, 0xc43cf2bb00, 0xc43cf2b200, 0xc44d294c60)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:1729 +0x1c93
created by github.com/dgraph-io/dgraph/query.ProcessGraph
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/query/query.go:1701 +0x18ea
```

--- a/query/query.go
+++ b/query/query.go
@@ -1643,7 +1643,9 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
 		// If we have a filter SubGraph which only contains an operator,
 		// it won't have any attribute to work on.
 		// This is to allow providing SrcUIDs to the filter children.
-		sg.DestUIDs = sg.SrcUIDs
+		// Each filter use it's own (shallow) copy of SrcUIDs, so there is no race conditions,
+		// when multiple filters replace their sg.DestUIDs
+		sg.DestUIDs = &protos.List{sg.SrcUIDs.Uids}
 	} else {
 		if len(sg.SrcFunc) > 0 && sg.SrcFunc[0] == "var" {
 			// If its a var() filter, we just have to intersect the SrcUIDs with DestUIDs

--- a/query/query_test.go
+++ b/query/query_test.go
@@ -3918,6 +3918,28 @@ func TestToFastJSONFilterNot3(t *testing.T) {
 		`{"me":[{"gender":"female","name":"Michonne","friend":[{"name":"Daryl Dixon"}]}]}`, js)
 }
 
+func TestToFastJSONFilterNot4(t *testing.T) {
+	populateGraph(t)
+	query := `
+		{
+			me(id:0x01) {
+				name
+				gender
+				friend (first:2) @filter(not anyofterms(name, "Andrea")
+				and not anyofterms(name, "glenn")
+				and not anyofterms(name, "rick")
+			) {
+					name
+				}
+			}
+		}
+	`
+
+	js := processToFastJSON(t, query)
+	require.JSONEq(t,
+		`{"me":[{"gender":"female","name":"Michonne","friend":[{"name":"Daryl Dixon"}]}]}`, js)
+}
+
 func TestToFastJSONFilterAnd(t *testing.T) {
 	populateGraph(t)
 	query := `

